cmd/api/src/api/middleware/compression.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package middleware
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/specterops/bloodhound/log"
+	"github.com/specterops/bloodhound/src/api"
+)
+
+type GzipResponseWriter struct {
+	http.ResponseWriter
+	gw *gzip.Writer
+}
+
+func NewGzipResponseWriter(w http.ResponseWriter) *GzipResponseWriter {
+	return &GzipResponseWriter{
+		ResponseWriter: w,
+		gw:             gzip.NewWriter(w),
+	}
+}
+
+func (s *GzipResponseWriter) Write(p []byte) (int, error) {
+	return s.gw.Write(p)
+}
+
+func (s *GzipResponseWriter) Close() error {
+	return s.gw.Close()
+}
+
+func CompressionMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
+		var (
+			gw  *GzipResponseWriter
+			err error
+		)
+
+		for headerName, headerValues := range request.Header {
```

Comment:
`request.Header` is a `map[string][]string` so there is no need to loop through the whole map to get the encoding values and what you have here might be a little overkill. Consider simplifying to something like:

``` go
if encodings := strings.Split(request.Header.Get(headers.ContentEncoding.String()), ","); len(encodings) > 0 {}
```

Reply:
`Header.Get()` won't work here because it will only return the first header value and ignore any others. I opted to use `Header.Values()`, which still requires joining and splitting again, but it has been collapsed.